pkg/clusterversion/cv.go

```diff
@@ -151,24 +178,37 @@ func (c *clusterVersionClient) HasUpgradeCompleted(cv *configv1.ClusterVersion,
 
 // isEqualVersion compare the upgrade version state for cv and uc
 func isEqualVersion(cv *configv1.ClusterVersion, uc *upgradev1alpha1.UpgradeConfig) bool {
-	if cv.Spec.DesiredUpdate != nil &&
-		cv.Spec.DesiredUpdate.Version == uc.Spec.Desired.Version {
-		return true
-	}
+	return cv.Spec.DesiredUpdate != nil && (cv.Spec.DesiredUpdate.Version == uc.Spec.Desired.Version)
+}
 
-	return false
+// isEqualImage compare the upgrade version state for cv and uc
+func isEqualImage(cv *configv1.ClusterVersion, uc *upgradev1alpha1.UpgradeConfig) bool {
+	return cv.Spec.DesiredUpdate != nil && (cv.Spec.DesiredUpdate.Image == uc.Spec.Desired.Image)
 }
 
-// hasUpgradeCommenced checks if the upgrade has commenced
+// HasUpgradeCommenced checks if the upgrade has commenced based on version or image
 func (c *clusterVersionClient) HasUpgradeCommenced(uc *upgradev1alpha1.UpgradeConfig) (bool, error) {
 
 	clusterVersion, err := c.GetClusterVersion()
 	if err != nil {
 		return false, err
 	}
 
-	if !isEqualVersion(clusterVersion, uc) {
-		return false, nil
+	if !empty(uc.Spec.Desired.Version) && !empty(uc.Spec.Desired.Channel) && empty(uc.Spec.Desired.Image) {
+		if !isEqualVersion(clusterVersion, uc) {
+			return false, nil
+		} else {
+			logger.Info(fmt.Sprintf("ClusterVersion is already set to Channel %s Version %s", uc.Spec.Desired.Channel, uc.Spec.Desired.Version))
+		}
+	}
+
+	// TODO: Remove version dependency after work on OSD-7609
+	if !empty(uc.Spec.Desired.Version) && !empty(uc.Spec.Desired.Image) && empty(uc.Spec.Desired.Channel) {
```

Comment:
what if all the version, channel and image specified?

Reply:
If all three mentioned that validation pkg will throw error in the beginning itself before proceeding with upgrade. However, this block would only focus on checking if upgrade already commenced if just image+version is specified. Do you think we still need that check in `clusterversion` package? I think that might be dead code which won't be reached at all because `validation` would handle it before.